File: typescript/src/clients/generations.ts
// =============================================================================
// Task API SDK - Generations Client
// =============================================================================

import { BaseClient } from "../client-base.js";
import type { RequestOptions, UsageInfo } from "../types.js";

// -----------------------------------------------------------------------------
// Generations-specific Types
// -----------------------------------------------------------------------------

/** A recorded generation (function call). */
export interface Generation {
  readonly id: string;
  readonly function_name: string;
  readonly input: unknown;
  readonly output: unknown;
  readonly created_at: string;
  readonly model?: string;
  readonly usage?: UsageInfo;
}

/** Pagination metadata for generation listings. */
export interface GenerationsListMeta {
  readonly page: number;
  readonly page_size: number;
  readonly total: number;
  readonly total_pages: number;
}

/** Response from listing generations. */
export interface GenerationsListResponse {
  readonly data: Generation[];
  readonly meta: GenerationsListMeta;
}

/** Parameters for listing generations. */
export interface ListGenerationsParams {
  /** Semantic search query (uses hybrid dense+sparse search). */
  query?: string;
  /** Page number (default 1). */
  page?: number;
  /** Items per page (default 50). */
  page_size?: number;
}

/** Response from deleting a generation. */
export interface DeleteGenerationResponse {
  readonly deleted: boolean;
}

// -----------------------------------------------------------------------------
// Generations Client
// -----------------------------------------------------------------------------

/**
 * Client for the Generations API endpoints.
 * Provides methods to list, get, and delete recorded HTTP request/response generations.
 */
export class GenerationsClient extends BaseClient {
  /**
   * List recorded HTTP request/response generations with pagination.
   *
   * @param params - Optional query parameters for filtering and pagination.
   * @param options - Optional request options.
   * @returns A paginated list of generations.
   */
  async listGenerations(
    params?: ListGenerationsParams,
    options?: RequestOptions,
  ): Promise<GenerationsListResponse> {
    return this.get<GenerationsListResponse>(
      "/v3/generations",
      {
        query: params?.query,
        page: params?.page,
        page_size: params?.page_size,
      },
      options,
    );
  }

  /**
   * Get a specific recorded generation by ID.
   *
   * @param id - The generation ID.
   * @param options - Optional request options.
   * @returns The recorded generation with request and response data.
   */
  async getGeneration(id: string, options?: RequestOptions): Promise<Generation> {
    return this.get<Generation>(`/v3/generations/${encodeURIComponent(id)}`, undefined, options);
  }

  /**
   * Delete a specific recorded generation.
   *
   * @param id - The generation ID.
   * @param options - Optional request options.
   * @returns Confirmation of deletion.
   */
  async deleteGeneration(id: string, options?: RequestOptions): Promise<DeleteGenerationResponse> {
    return this.delete<DeleteGenerationResponse>(
      `/v3/generations/${encodeURIComponent(id)}`,
      options,
    );
  }
}

File: python/src/opperai/clients/generations.py
"""Opper SDK — Generations Client."""

from __future__ import annotations

from typing import Any
from urllib.parse import quote

from .._base_client import BaseClient
from ..types import GenerationsListMeta, GenerationsListResponse, RequestOptions, _from_dict


class GenerationsClient:
    """Client for the Generations API endpoints."""

    def __init__(self, client: BaseClient) -> None:
        self._client = client

    def list(
        self,
        *,
        query: str | None = None,
        page: int | None = None,
        page_size: int | None = None,
        options: RequestOptions | None = None,
    ) -> GenerationsListResponse:
        params: dict[str, Any] = {}
        if query is not None:
            params["query"] = query
        if page is not None:
            params["page"] = page
        if page_size is not None:
            params["page_size"] = page_size
        data = self._client._get("/v3/generations", query=params, options=options)
        meta_raw = (data or {}).get("meta", {})
        return GenerationsListResponse(
            data=(data or {}).get("data", []),
            meta=_from_dict(GenerationsListMeta, meta_raw) if meta_raw else GenerationsListMeta(),
        )

    async def list_async(
        self,
        *,
        query: str | None = None,
        page: int | None = None,
        page_size: int | None = None,
        options: RequestOptions | None = None,
    ) -> GenerationsListResponse:
        params: dict[str, Any] = {}
        if query is not None:
            params["query"] = query
        if page is not None:
            params["page"] = page
        if page_size is not None:
            params["page_size"] = page_size
        data = await self._client._get_async("/v3/generations", query=params, options=options)
        meta_raw = (data or {}).get("meta", {})
        return GenerationsListResponse(
            data=(data or {}).get("data", []),
            meta=_from_dict(GenerationsListMeta, meta_raw) if meta_raw else GenerationsListMeta(),
        )

    def get(self, id: str, *, options: RequestOptions | None = None) -> dict[str, Any]:
        return self._client._get(f"/v3/generations/{quote(id, safe='')}", options=options)

    async def get_async(self, id: str, *, options: RequestOptions | None = None) -> dict[str, Any]:
        return await self._client._get_async(f"/v3/generations/{quote(id, safe='')}", options=options)

    def delete(self, id: str, *, options: RequestOptions | None = None) -> dict[str, Any]:
        return self._client._delete(f"/v3/generations/{quote(id, safe='')}", options=options)

    async def delete_async(self, id: str, *, options: RequestOptions | None = None) -> dict[str, Any]:
        return await self._client._delete_async(f"/v3/generations/{quote(id, safe='')}", options=options)
